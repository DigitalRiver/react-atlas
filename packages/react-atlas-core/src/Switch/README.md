Switch component allows to define a particular-styled checkbox.

###### Default switch:
    initialState = {checked: true}
    _handleChange = () => {
        setState({checked: !state.checked})
    };
    
    <Switch onChange={_handleChange} checked={state.checked}/>

###### Different sized switches:
    initialState = {
        switch1: false,
        switch2: false,
        switch3: false
    }
    _handleChange = (e) => {
        setState({[`${e.target.name}`]: !state[e.target.name]})
    };
    <div>
      <Switch name="switch1" small onChange={_handleChange} checked={state.switch1}/>
      <Switch name="switch2" medium onChange={_handleChange} checked={state.switch2}/>
      <Switch name="switch3" large onChange={_handleChange} checked={state.switch3}/>
    </div>

###### Multiple switches tied to same behavior (name prop):
    initialState = {
        switch1: false,
        switch2: false,
        switch3: false
    }
    _handleChange = (e) => {
        setState({[`${e.target.id}`]: !state[e.target.id]})
    };
    <div>
      <Switch id="switch1" name="customFeature" onChange={_handleChange} checked={state.switch1}/>
      <Switch id="switch2" name="customFeature" onChange={_handleChange} checked={state.switch2}/>
      <Switch id="switch3" name="customFeature" onChange={_handleChange} checked={state.switch3}/>
    </div>

###### Set other colors for on/off states (hex, rgb, css-valid color name):

    initialState = {
        switch1: false,
        switch2: false
    }
    _handleChange = (e) => {
        setState({[`${e.target.name}`]: !state[e.target.name]})
    };
    <div>
      <Switch name="switch1" onColor="#00bbaa" offColor="#000000" onChange={_handleChange} checked={state.switch1}/>
      <Switch name="switch2" onColor="green" offColor="black" onChange={_handleChange} checked={state.switch2}/>
    </div>

###### Disabled or hidden switch:

    <div>
      <Switch disabled/>
      <Switch hidden/>
    </div>

###### Execute method when onChange event occurs:

    initialState = {
        switch1: false
    }
    customSwitchChange = () => {
    	console.info("onChange event!");
    }
    
    _handleChange = (e) => {
        setState({[`${e.target.name}`]: !state[e.target.name]})
        customSwitchChange();
    };
    
    <Switch name="switch1" onChange={_handleChange} checked={state.switch1}/>

###### Make checks before changing component's state (onbeforechange). It receives a value with current checked state:

    initialState = {
        switch1: false
    }
    _customSwitchBeforeChange = (e) => {
        let accept = confirm("Do you want to check this?");
        if(accept){
            setState({[`${e.target.name}`]: !state[e.target.name]})
        }
    };
    
    <Switch name="switch1" onBeforeChange={_customSwitchBeforeChange} checked={state.switch1}/>

This method should return true or false, allowing or not the onChange handler to be fired. If undefined, will prevent it.

###### Update switch props

    initialState = {checked: true}
    _handleChangeProps = () => {
      setState({ size: 'small' });
    };
    
    _handleChange = () => {
        setState({checked: !state.checked})
    };
    let switchComp = state.size === 'small'? 
        <Switch small onChange={_handleChange} checked={state.checked}/>:
        <Switch onChange={_handleChange} checked={state.checked}/>;
    <div>
        <Button primary onClick={_handleChangeProps}>Change Props</Button>
        {switchComp}
    </div>
