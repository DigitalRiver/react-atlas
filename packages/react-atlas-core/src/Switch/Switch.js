import React from "react";
import PropTypes from "prop-types";
import { InputCore } from "../Input";
import cx from "classnames";

class Switch extends React.PureComponent {
  constructor(props) {
    super(props);
    // Initial state
    this.state = {
      "checked": this.props.checked || false,
      "disabled": this.props.disabled || false,
      "valid": true
    };

    /* Classes and styles setup */
    let small = this.props.small && !this.props.medium && !this.props.large,
      medium = !this.props.small && !this.props.large,
      large = this.props.large && !this.props.medium && !this.props.small,
      onColor = this.props.onColor,
      buttonColor = this.props.buttonColor;

    let offClassName = cx({
      "sliderSmall": small,
      "sliderMedium": medium,
      "sliderLarge": large,
      "sliderEnabledOffColor": !this.state.disabled,
      "sliderDisabledOffColor": this.state.disabled
    });

    let inputClassName = cx({
      "inputSmall": small,
      "inputMedium": medium,
      "inputLarge": large
    });

    let buttonClassName = cx({
      "handleSmall": small,
      "handleMedium": medium,
      "handleLarge": large,
      "handleDisabledColor": this.state.disabled && !this.props.checked ,
      "handleDisabledCheckedColor" : this.state.disabled && this.props.checked,
      "handleEnabledColor": !this.state.disabled
    });

    let onClassName = cx({
      "onColorSmall": small,
      "onColorMedium": medium,
      "onColorLarge": large,
      "sliderDisabledOnColor": this.state.disabled
    });

    let onColorStyle = {
      "background": onColor
    };

    let buttonColorStyle = {
      "background": buttonColor
    };

    this.classes = {
      inputClassName,
      buttonClassName,
      offClassName,
      onClassName
    };

    this.styles = {
      onColorStyle,
      buttonColorStyle
    };
  }

  componentWillReceiveProps = (nextProps) =>  {
      if(nextProps.checked !== this.props.checked) {
        this.setState({ "checked": !this.state.checked })
      }
      if(nextProps.disabled !== this.props.disabled) {
        this.setState({ "disabled": !this.state.disabled })
      }
  };

  // Handles new checkbox clicks and sets value and checked status of hidden input
  _clickHandler = () => {
    if (!this.state.disabled) {
      if (typeof this.props.onBeforeChange !== "undefined") {
        let result = this.props.onBeforeChange(this.state.checked);
        if(result === false) {
          return;
        }
      }

      this.setState({ "checked": !this.state.checked }, function() {
        /* Check if onClick has been passed, if so call it. */
        if (typeof this.props.onClick !== "undefined") {
          this.props.onClick(
            this.state.checked,
            this.state.disabled
          );
        }
        /* Check if onChange has been passed, if so call it. */
        if (typeof this.props.onChange !== "undefined") {
          this.props.onChange(
            this.state.checked,
            this.state.disabled
          );
        }
      });
    }
  };

  render() {
    const {
      label,
      leftLabel,
      className, 
      name, 
      hidden, 
      style,
      inline,
      id
    } = this.props;

    const classes = this.classes;
    const styles = this.styles;
    const forId = id !== "" && name !== "" ? id : "";

    let switchWrapperClasses = cx(
        {
          "switch": true
        }
    );

    let switchClasses = cx(
      {
        "disabled": this.state.disabled,
        hidden,
        inline,
        "leftLabelContent": leftLabel
      },
      classes.offClassName
    );

    let labelClasses = cx(
        {
            "leftLabel": leftLabel,
            "label":true,
            "labelFont":true,
            "labelSpacing":true
        }
    );

    let switchLabel = label &&
      <div styleName={labelClasses}>
        <label htmlFor={forId}>
          <span>{label}</span>
        </label>
      </div>
    ;
    return (
      <div
        className={cx(className)}
        styleName={switchWrapperClasses}
      >
        {switchLabel}
        <div
          onClick={this._clickHandler}
          styleName={switchClasses}
          style={style}
        >
          <InputCore
            type="checkbox"
            name={name}
            id={id}
            styleName={classes.inputClassName}
            checked={this.state.checked}
            disabled={this.state.disabled}
            hidden={this.state.disabled}
          />
          {/* handle */}
          <div
            styleName={classes.buttonClassName}
            style={styles.buttonColorStyle}
          />
          {/* background */}
          <div
            styleName={classes.onClassName}
            style={styles.onColorStyle} />
        </div>
      </div>
    );
  }
}

Switch.propTypes = {
  /**
   * Text for checkbox label
   * @examples 'Some Label'
   */
  "label": PropTypes.string,
  /**
   * Allows user to move the label to the left of the Switch instead of above it
   */
  "leftLabel": PropTypes.bool,
  /** An Object, array, or string of CSS classes to apply to Switch.*/
  "className": PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array
  ]),
  /**
   * Determines if the switch input is disabled.
   * @examples '<Switch disabled={condition}/>'
   */
  "disabled": PropTypes.bool,
  /**
   * Determines if the switch input is hidden.
   * @examples '<Switch hidden={condition}/>'
   */
  "hidden": PropTypes.bool,
  /**
   * Determines if the switch input is checked.
   * @examples '<Switch checked={condition}/>'
   */
  "checked": PropTypes.bool,
  /**
   * Define a name for the switch input.
   * @examples '<Switch name="test"/>'
   */
  "name": PropTypes.string,
  /**
   * Defines an id for the switch input.
   * @examples '<Input type="text" name="test"/>'
   */
  "id": PropTypes.string,
  /**
   * Sets color that will be displayed when the switch is checked.
   * @examples '<Switch onColor="#ababab"/>'
   */
  "onColor": PropTypes.string,
  /**
   * Allows user to pass a callback for click events.
   */
  "onClick": PropTypes.func,
  /**
   * Sets color that will be displayed when the switch is unchecked.
   * @examples '<Switch offColor="#d3d3d3"/>'
   */
  "offColor": PropTypes.string,
  /**
   * Sets color that will be displayed for the inner button.
   * @examples '<Switch buttonColor="#ffffff"/>'
   */
  "buttonColor": PropTypes.string,
  /**
   * Defines a small sized switch element.
   * @examples '<Switch small/>'
   */
  "small": PropTypes.bool,
  /**
   * Defines a medium sized switch element.
   * @examples '<Switch medium/>'
   */
  "medium": PropTypes.bool,
  /**
   * Defines a large sized switch element.
   * @examples '<Switch large/>'
   */
  "large": PropTypes.bool,
  /**
   * Sets a handler function to be executed when onChange event occurs (at input element).
   * @examples <Switch onChange={this.customOnChangeFunc}/>
   */
  "onChange": PropTypes.func,
  /**
   * Sets a handler function to be executed before the onChange event occurs.
   * @examples <Switch onBeforeChange={this.customOnBeforeChangeFunc}/>
   */
  "onBeforeChange": PropTypes.func,
  /**
   * Sets the Switch as an inline-block element
   */
  "inline": PropTypes.bool,

  /**
   * Pass inline styling here.
   */
  "style": PropTypes.object
};

Switch.defaultProps = {
  "checked": false,
  "className": "",
  "disabled": false,
  "hidden": false
};

export default Switch;
