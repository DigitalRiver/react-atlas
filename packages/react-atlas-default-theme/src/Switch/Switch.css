/*
  background sizes
*/

.slider {
  composes: cursor-pointer from '../styles.css';
  composes: z1 from '../styles.css';
  overflow:hidden;
  display: block;
  position:relative;
  margin: 0 .25rem .25rem 0;
}

.sliderSmall{
  composes: slider;
  width: 50px;
  height: 28px;
  border-radius: 14px;
}

.sliderMedium{
  composes: slider;
  width: 75px;
  height: 40px;
  border-radius: 20px;
}

.sliderLarge{
  composes: slider;
  width: 100px;
  height: 55px;
  border-radius: 27.5px;
}

/*
  background color states
*/

.sliderDisabledOffColor {
  composes: bg-silver from '../styles.css';
}

.sliderDisabledOnColor {
  composes: bg-silver from '../styles.css';
}

.sliderEnabledOffColor {
  composes: bg-gray from '../styles.css';
}

/*
  the handle
*/
.handle {
  composes: circle from '../styles.css';
  display:block;
  position:absolute;
  transition:left 0.2s;
}

.handleSmall{
  composes: handle;
  top: 4px;
  left: 4px;
  width: 20px;
  height: 20px;
}

.handleMedium{
  composes: handle;
  top: 5px;
  left: 5px;
  width: 30px;
  height: 30px;
}

.handleLarge{
  composes: handle;
  top: 5px;
  left: 5px;
  width: 45px;
  height: 45px;
}

/*
  handle color states
*/

.handleDisabledColor{
  composes: bg-gray from '../styles.css';
}

.handleEnabledColor{
  composes: bg-white from '../styles.css';
}

.input {
    display:none;
}

/*
  Distance the button translates left to right
*/

.inputSmall:checked + .handle{
    left:25px;
}
.inputMedium:checked + .handle{
    left:40px;
}
.inputLarge:checked + .handle{
    left:50px;
}

/*
  On color sizes
*/

.onColor{
  composes: bg-primary from '../styles.css';
  position: relative;
  composes: zNeg from '../styles.css';
  left: -48px;
}

.onColorSmall{
  composes: onColor;
  width: 19px;
  height: 28px;
}

.onColorMedium{
  composes: onColor;
  width: 32px;
  height: 40px;
}

.onColorLarge{
  composes: onColor;
  width: 48px;
  height: 55px;
}

.input:checked ~ .onColorSmall{
    width: 100px;
    transition: 0.2s;
}

.input:checked ~ .onColorMedium{
    width: 132px;
    transition: 0.2s;
}

.input:checked ~ .onColorLarge{
    width: 158px;
    transition: 0.2s;
}

.disabled{
  text-decoration: none;
  composes: cursor-not-allowed from '../styles.css';
  pointer-events: none;
}

.hidden{
  display: none;
  visibility: hidden;
}

.inputSmall{
  composes: input;
}

.inputMedium{
  composes: input;
}

.inputLarge{
  composes: input;
}

.inline {
  display: inline-block;
  position: relative;
  top: 16px;
}

