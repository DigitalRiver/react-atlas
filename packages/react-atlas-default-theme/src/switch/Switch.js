import { Switch } from "react-atlas-core";
import style from "./switch.css";
import createComponent from "../createComponent";

export default createComponent(Switch, style);
