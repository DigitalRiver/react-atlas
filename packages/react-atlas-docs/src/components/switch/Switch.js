import React, { PropTypes } from "react";
import * as Atlas from 'react-atlas';

export default class Switch extends React.Component {
  constructor(props){
    super(props)
  }
  render() {
    return (
      <Atlas.Switch {...this.props}></Atlas.Switch>
    )
  }
}