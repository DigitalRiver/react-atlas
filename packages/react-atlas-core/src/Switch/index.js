import { default as SwitchCore } from "./Switch";
export { SwitchCore };
