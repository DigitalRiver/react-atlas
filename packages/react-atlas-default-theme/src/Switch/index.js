import * as SwitchStyle from "./Switch.css";
export { SwitchStyle };
