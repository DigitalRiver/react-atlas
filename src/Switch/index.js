import Switch from "./Switch.js";
export { Switch };
