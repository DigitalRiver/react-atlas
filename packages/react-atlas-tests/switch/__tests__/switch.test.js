import React from "react";
import { mount } from "enzyme";
import { SwitchCore } from "../../../react-atlas-core/src/Switch/index";

describe("Testing switch component", () => {
  it("Switch component - Basic test", function() {
    const comp = mount(<SwitchCore onColor="black" />);
  });

  it("Switch component - Basic test one click switch", function() {
    const comp = mount(<SwitchCore onColor="black" />);
    expect(comp.state().checked).toEqual(false);
    comp.find("Input").simulate("click");
    expect(comp.state().checked).toEqual(true);
  });

  it("Switch component - Basic test two clicks switch", function() {
    const comp = mount(<SwitchCore onColor="black" />);
    expect(comp.state().checked).toEqual(false);
    comp.find("Input").simulate("click");
    expect(comp.state().checked).toEqual(true);
    comp.find("Input").simulate("click");
    expect(comp.state().checked).toEqual(false);
  });

  it("Switch component - Basic test one click switch with onBeforeChange", function() {
    const comp = mount(
      <SwitchCore
        onColor="black"
        onBeforeChange={function() {
          console.log("onBeforeChange triggered");
          return true;
        }}
      />
    );
    expect(comp.state().checked).toEqual(false);
    comp.find("Input").simulate("click");
    expect(comp.state().checked).toEqual(true);
  });

  it("Switch component - Basic test one click switch with onBeforeChange II", function() {
    const comp = mount(
      <SwitchCore
        onColor="black"
        onBeforeChange={function() {
          console.log("onBeforeChange triggered");
          return false;
        }}
      />
    );
    expect(comp.state().checked).toEqual(false);
    comp.find("Input").simulate("click");
    expect(comp.state().checked).toEqual(false);
  });
});
