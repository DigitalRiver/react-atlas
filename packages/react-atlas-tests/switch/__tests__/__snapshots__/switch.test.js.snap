// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test switch component render Render correctly 1`] = `
<div
  className="Class"
  styleName="switch"
>
  <div
    onClick={[Function]}
    style={undefined}
    styleName="sliderMedium sliderEnabledOffColor"
  >
    <input
      checked={false}
      className=""
      id="ID"
      name="Name"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyPress={[Function]}
      onPaste={[Function]}
      style={undefined}
      type="checkbox"
    />
    <div
      style={
        Object {
          "background": "grey",
        }
      }
      styleName="handleMedium handleEnabledColor"
    />
    <div
      style={
        Object {
          "background": "black",
        }
      }
      styleName="onColorMedium"
    />
  </div>
</div>
`;
